### PR TITLE
Fix test_decorator issue #269

### PR DESCRIPTION
**Pull request by zhiruiluo**
To fix issue #269.

### CODE REVIEW

**Comment by zhiruiluo** [comment]
This doesn't exactly behave like the original partial function via changing defaults.

**Comment by zhiruiluo** [comment]
Added a ```delay_evaluation_wrapper``` to prevent collection error.

File: test/test_decorator.py [hunk 1]
```diff
@@ -69,29 +75,41 @@ def _wrapper(*other_args, **other_kwargs):
...
+
+
 @pytest.mark.parametrize(
-    "args, expected, fn",
+    "args, expected, delay_wrapper",
     [
         ("", 1, partial(_fn_with_positional_only, 1)),
         ("2", 2, partial(_fn_with_positional_only, 1)),
-        ("2", 2, _fn_with_positional_only),
+        ("2", 2, partial(_fn_with_positional_only)),
```
Comment (lebrice) [inline test/test_decorator.py:94]: Hey there @zhiruiluo , Would you mind explaining a few things:
- why is the `delay_wrapper` is necessary here?
- Did the `inspect` module change from 3.11 to 3.11.3? Is that what broke this test?

Comment (zhiruiluo) [reply test/test_decorator.py:94]: The recently PR https://github.com/python/cpython/pull/103557 disallows pos-or-kw params without default after pos-only with default, which breaks the test for Python 3.11.4 released on June 6.

Comment (zhiruiluo) [reply test/test_decorator.py:94]: They will likely backport this change to all Python 3.11 verions.

File: test/test_decorator.py [hunk 2]
```diff
@@ -43,6 +41,14 @@ def pop(self, index, default_value=None):
...
 
 
+def delay_evaluation_wrapper(fn: Callable) -> Callable:
+    def delayed_call(*args, **kwargs):
+        return lambda: fn(*args, **kwargs)
+
+    return delayed_call
+
+
+@delay_evaluation_wrapper
```
Comment (lebrice) [inline test/test_decorator.py:51]:
```suggestion
@functools.partial
```

**Review by lebrice — CHANGES_REQUESTED** [review]
I don't really understand the change in the inspect module, and I also don't quite understand your changes here either.. 

I do understand that you want to delay the call to `change_signature(fn_with_all_argument_types, 1, c=1)` because executing it in Python 3.11.4 raises a `ValueError("non-default argument follows default argument")`.

However, before your change, `partial` here is meant as `change_signature`, it makes a variant of a function with a different signature (different default values in the signature).

Here your `delayed_wrapper` is exactly the same as `functools.partial`, and when applied to the `partial` (a.k.a. `change_signature`, it doesn't change anything, as far as I can tell. Why are you applying it to the `change_signature` function?

It also seems to my eyes like you are changing the behaviour of the rest of the test cases by only adding this additional level of "delay" to only _some_ of the test cases.


I'm going to re-read your PR, but I'd really appreciate if you could clarify this a bit for me.

File: test/test_decorator.py (hunk 2 above)
Comment (lebrice) [reply test/test_decorator.py:51]: And at that point, as far as I can tell, this doesn't do anything, so why are we adding it?

**Comment by lebrice** [comment]
Closed in favour of #273

**Pull request closed without merging**